src/x11.cc

```diff
@@ -1205,61 +1084,83 @@ void print_desktop_name(struct text_object *obj, char *p,
 }
 
 #ifdef OWN_WINDOW
+enum class x11_strut : size_t {
+  LEFT,
+  RIGHT,
+  TOP,
+  BOTTOM,
+  LEFT_START_Y,
+  LEFT_END_Y,
+  RIGHT_START_Y,
+  RIGHT_END_Y,
+  TOP_START_X,
+  TOP_END_X,
+  BOTTOM_START_X,
+  BOTTOM_END_X,
+  COUNT
+};
+constexpr size_t operator*(x11_strut index) {
+  return static_cast<size_t>(index);
+}
+
 /* reserve window manager space */
-void set_struts(int sidenum) {
-  Atom strut;
-  if ((strut = ATOM(_NET_WM_STRUT)) != None) {
+void set_struts(alignment align) {
+  // Middle and none align don't have least significant bit set.
+  // Ensures either vertical or horizontal axis are start/end
+  if ((*align & 0b0101) == 0) return;
+
+  Atom strut = ATOM(_NET_WM_STRUT);
+  if (strut != None) {
     /* reserve space at left, right, top, bottom */
-    signed long sizes[12] = {0};
+    uint32_t sizes[*x11_strut::COUNT] = {0};
     int i;
 
-    /* define strut depth */
-    switch (sidenum) {
-      case 0:
-        /* left side */
-        sizes[0] = window.x + window.width;
-        break;
-      case 1:
-        /* right side */
-        sizes[1] = display_width - window.x;
-        break;
-      case 2:
-        /* top side */
-        sizes[2] = window.y + window.height;
-        break;
-      case 3:
-        /* bottom side */
-        sizes[3] = display_height - window.y;
+    switch (vertical_alignment(align)) {
+      case axis_align::START:
+        sizes[*x11_strut::TOP] =
+            std::min(window.y + window.height, display_height);
+        sizes[*x11_strut::TOP_START_X] = window.x;
+        sizes[*x11_strut::TOP_END_X] =
+            std::min(window.x + window.width, display_width);
+      case axis_align::END:
+        sizes[*x11_strut::BOTTOM] =
+            window.y < display_height ? display_height - window.y : 0;
+        sizes[*x11_strut::BOTTOM_START_X] = window.x;
+        sizes[*x11_strut::BOTTOM_END_X] =
+            std::min(window.x + window.width, display_width);
+      case axis_align::MIDDLE:
+        // can't reserve space in middle of the screen
+      default:
         break;
     }
-
-    /* define partial strut length */
-    if (sidenum <= 1) {
-      sizes[4 + (sidenum * 2)] = window.y;
-      sizes[5 + (sidenum * 2)] = window.y + window.height;
-    } else if (sidenum <= 3) {
-      sizes[4 + (sidenum * 2)] = window.x;
-      sizes[5 + (sidenum * 2)] = window.x + window.width;
-    }
-
-    /* check constraints */
-    for (i = 0; i < 12; i++) {
-      if (sizes[i] < 0) {
-        sizes[i] = 0;
-      } else {
-        if (i <= 1 || i >= 8) {
-          if (sizes[i] > display_width) { sizes[i] = display_width; }
-        } else {
-          if (sizes[i] > display_height) { sizes[i] = display_height; }
-        }
-      }
```

Comment:
We can also skip this loop (12 jumps) entirely because we've already handled constraints for each side separately.